File: leetcode/BackTracking/17. 电话号码对应的字母组合(combination).go
package BackTracking

func letterCombinations(digits string) []string {
	if len(digits) == 0{
		return []string(nil)
	}

	hashmap := map[string]string{
		"2": "abc",
		"3": "def",
		"4": "ghi",
		"5": "jkl",
		"6": "mno",
		"7": "pqrs",
		"8": "tuv",
		"9": "wxyz",}

	var res []string
	var dfs func(cur int, temp string)

	dfs = func(cur int, temp string) {

		if len(temp) == len(digits) {
			// 字符串不能像数组一样
			// 使用 append([]int(nil), temp...)
			res = append(res, temp)
			return
		}

		// 通过string()就可以把byte转化成string
		curdigit := string(digits[cur])
		curbytes := hashmap[curdigit]
		for i:= 0; i < len(curbytes); i ++ {
			dfs(cur + 1, temp + string(curbytes[i]))
		}

	}

	dfs(0, "")
	return res


}

File: lc/BackTracking(DFS)/17. 电话号码对应的字母组合(combination).py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if digits == "":
            return []

        self.nm = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz"
        }
        self.res = []
        self.temp = []
        self.digits = digits
        self.n
        self.helper(0)

        return self.res

    def helper(self, startFrom):
        if len(self.temp) > self.n:
            return

        if len(self.temp) == self.n:
            self.res.append("".join(self.temp[:]))
            return

        for i in range(startFrom, self.n):
            options = self.nm[self.digits[i]]
            for j in options:
                self.temp.append(j)
                self.helper(i + 1)
                self.temp.pop()


"""
与77题相似
"""